### ISSUE TITLE
Emit last exit code to pipeline variable to allow for use in task conditions

### ISSUE THREAD

**Issue by Justin-DynamicD** [charleszipp/azure-pipelines-tasks-terraform#68]
Currently, while I can define `commandOptions` when doing a Terraform plan, if I use `-detailed-exit-code` the process will fail as a non-empty diff will result in ExitCode 2.

Maybe you already solved this, but if not, are there plans to possibly add a detailed exit code switch that returns a var we can use as a conditional in later steps?

**Comment by charleszipp** [comment]
> Currently, while I can define `commandOptions` when doing a Terraform plan, if I use `-detailed-exit-code` the process will fail as a non-empty diff will result in ExitCode 2.

@Justin-DynamicD, this is a known issue captured in #16. If you can explain why you are needing to use -detailed-exit-code that would help me prioritize the fix for this issue.

> Maybe you already solved this, but if not, are there plans to possibly add a detailed exit code switch that returns a var we can use as a conditional in later steps?

Can you elaborate a bit on the use case here? What would the var contain? Would this be emitted as pipeline var? How would you plan to make use of this feature? I believe what you are asking about has been requested by a few others. However, I would like to understand your case in more detail to make sure its the same as others have requested.

**Comment by Justin-DynamicD** [comment]
The use case in my scenario is fairly simple: it allows me to skip the apply step if the dif comes back 0 using a simple condition on the yaml task.  Right now, I basically just run a custom powershell task to get around it and use your code to do the init (as it does a great job of passing through the extra var for my partial config):

```yaml
- task: PowerShell@2
        displayName: Terraform Plan
        condition: succeeded()
        inputs:
          targetType: inline
          workingDirectory: $(Build.SourcesDirectory)/${{ parameters['WorkingDirectory'] }}
          script: |
            $ErrorActionPreference = "SilentlyContinue"
            terraform plan -detailed-exitcode
            if ($LASTEXITCODE -eq 2 ) {
              $LASTEXITCODE = 0
              Write-Output "Apply is required"
              Write-Output "##vso[task.setvariable variable=infraUpdate;]$true"
              Write-Output "##vso[task.setvariable variable=infraUpdate;isOutput=true]$true"
            }
            else {
              Write-Output "No changes needed"
              Write-Output "##vso[task.setvariable variable=infraUpdate;]$false"
              Write-Output "##vso[task.setvariable variable=infraUpdate;isOutput=true]$false"
            }
            exit $LASTEXITCODE

      - task: charleszipp.azure-pipelines-tasks-terraform.azure-pipelines-tasks-terraform-cli.TerraformCLI@0
        displayName: 'terraform apply'
        condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'), eq(variables['infraUpdate'], 'True'))
[...]
```

As you can see, I spend more lines dealing with the exit code than running terraform which is why I made the request.

The bigger reason I care about saving those few seconds is because my pipeline is somewhat modeled after the pattern outlined by [terragrunt](https://blog.gruntwork.io/terragrunt-how-to-keep-your-terraform-code-dry-and-maintainable-f61ae06959d8): there is a singular repo that has multiple Terraform environments in it, which can cause an unnecessary number of apply steps to run on every commit.  By simply capturing the exit code of the plan step, I can skip unnecessary steps and keep that apply time down (In this case I'm also making sure it doesn't run against a PR, but you get the idea).


Hopefully this makes sense.

**Comment by charleszipp** [comment]
@Justin-DynamicD thanks for providing the detailed use case and context. This makes complete sense and should be relatively trivial to implement. The node.js library for the pipeline tasks already returns the exit code so i have it in a variable before the task completes. Additionally, the library has a method for setting a variable. Assuming these two things work as I expect they do, this should be easy to implement.

One last question, you will need the detailed exit code issue resolved as well in order to replace the scripted step with terraform plan from my task correct?

**Comment by Justin-DynamicD** [comment]
In the context of how you renamed the request, yes :)  It becomes two asks , I guess:
- allow exitcode `2` to pass instead of fail ( issue #16)
- use output variables to capture exit code for later tasks. (new)

I think for simplicity/flexibility passing the raw exit code would be fine; I can easily update a condition for that and it's a little more consistent with what I imagine others would expect to see output.

Heck you could go a step further on #16 and add an optional list parameter for "valid" exit codes, defaulting to [0,2].  This way if Hashicorp ever introduces new exit codes the users have the ability to apply a custom valid range without waiting on a task update.  Your task will accept anything in the list, and always return the specific code in the output so the pipeline can take any appropriate action.  Maybe I'm getting too fancy in that case ... but it seems like it would work.

**Comment by charleszipp** [comment]
@Justin-DynamicD see #72. This PR will fix #16. I will open a second PR for emitting exit code to pipeline variable.

**Comment by charleszipp** [comment]
@Justin-DynamicD  see #74. This will produce the pipeline variables you need to determine if apply should be run.

**Comment by charleszipp** [comment]
released with 0.4.10
